jobs: some API cleanup

RaduBerinde opened the pull request:
#### jobs: clean up the hook interface

The resume hook interface is very awkward: each hook is called and
passed the job type, and only the right hook will work. This is
confusing and unnecessary - we can associate each hook with a specific
job type.

Also renaming "hook" to "constructor" which is more descriptive.

Part of #34418.

Release note: None

#### jobs: clean up the Resumer interface

An instance of Resumer is associated with a specific job. However, the
API looks like that of a singleton "manager" in that every function
gets the job as an argument. This is confusing when trying to
understand these APIs without any prior knowledge.

This change cleans this up by passing the job to the Resumer
constructors instead; implementations can store the job inside each
instance.

Part of #34418.

Release note: None

RaduBerinde committed:
jobs: clean up the hook interface

The resume hook interface is very awkward: each hook is called and
passed the job type, and only the right hook will work. This is
confusing and unnecessary - we can associate each hook with a specific
job type.

Also renaming "hook" to "constructor" which is more descriptive.

Part of #34418.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36354)
<!-- Reviewable:end -->

RaduBerinde committed:
jobs: clean up the Resumer interface

An instance of Resumer is associated with a specific job. However, the
API looks like that of a singleton "manager" in that every function
gets the job as an argument. This is confusing when trying to
understand these APIs without any prior knowledge.

This change cleans this up by passing the job to the Resumer
constructors instead; implementations can store the job inside each
instance.

Part of #34418.

Release note: None

danhhz commented:
nice! I've been meaning to do something like this for a while. I didn't look closely since Matt had already reviewed, but approach lgtm. Thanks for the cleanup!

RaduBerinde commented:
TFTRs!

bors r+
